rs/sns/cli/src/upgrade_sns_controlled_canister.rs

```diff
@@ -0,0 +1,609 @@
+            let proposal_str = printing::pretty(&proposal).unwrap();
+            println!("{proposal_str}");
+        }
+    } else {
+        unimplemented!(
+            "Direct canister upgrades are not implemented yet. Please use DFX:\n{}",
+            suggested_install_command(&wasm.path(), &candid_arg)
+        );
+    }
+
+    Ok(())
+}
+
+pub type BlockIndex = Nat;
+
+#[derive(CandidType, Deserialize, Debug, Clone)]
+pub enum CreateCanisterError {
+    InsufficientFunds {
+        balance: Nat,
+    },
+    TooOld,
+    CreatedInFuture {
+        ledger_time: u64,
+    },
+    TemporarilyUnavailable,
+    Duplicate {
+        duplicate_of: Nat,
+        // If the original transaction created a canister then this field will contain the canister id.
+        canister_id: Option<Principal>,
+    },
+    FailedToCreate {
+        fee_block: Option<BlockIndex>,
+        refund_block: Option<BlockIndex>,
+        error: String,
+    },
+    GenericError {
+        message: String,
+        error_code: Nat,
+    },
+}
+
+// ```candid
+// type CreateCanisterArgs = record {
+//     from_subaccount : opt vec nat8;
+//     created_at_time : opt nat64;
+//     amount : nat;
+//     creation_args : opt CmcCreateCanisterArgs;
+// };
+// ```
+#[derive(Clone, Eq, PartialEq, Debug, CandidType, Deserialize)]
+pub struct CreateCanisterArgs {
+    from_subaccount: Option<Vec<u8>>,
+    created_at_time: Option<u64>,
+    amount: Nat,
+    creation_args: Option<CreateCanister>,
+}
+
+// ```candid
+// type CreateCanisterSuccess = record {
+//     block_id : BlockIndex;
+//     canister_id : principal;
+// };
+// ```
+#[derive(Clone, Eq, PartialEq, Debug, CandidType, Deserialize)]
+pub struct CreateCanisterSuccess {
+    block_id: BlockIndex,
+    canister_id: PrincipalId,
+}
+
+impl Request for CreateCanisterArgs {
+    fn method(&self) -> &'static str {
+        "create_canister"
+    }
+
+    fn update(&self) -> bool {
+        true
+    }
+
+    fn payload(&self) -> std::result::Result<Vec<u8>, candid::Error> {
+        Encode!(self)
+    }
+
+    type Response = Result<CreateCanisterSuccess, CreateCanisterError>;
+}
+
+pub async fn cycles_ledger_create_canister<C: CallCanisters>(
+    agent: &C,
+    cycles_amount: u128,
+    subnet_selection: Option<SubnetSelection>,
+    settings: Option<CanisterSettingsArgs>,
+) -> Result<CreateCanisterSuccess, CreateCanisterError> {
+    let request = CreateCanisterArgs {
+        from_subaccount: None,
+        created_at_time: None,
+        amount: Nat::from(cycles_amount),
+        creation_args: Some(CreateCanister {
+            subnet_selection,
+            settings,
+            ..Default::default()
+        }),
+    };
+    agent
+        .call(CYCLES_LEDGER_CANISTER_ID, request)
+        .await
+        .expect("Cannot create canister")
+}
+
+fn format_full_hash(hash: &[u8]) -> String {
+    hash.iter()
+        .map(|b| format!("{:02x}", b))
+        .collect::<Vec<_>>()
+        .join("")
+}
```

Comment:
Seems silly to do this in here and in ic-nervous-system-agent (and ic-sns-governance and ic-nervous-system-integration-tests). Why not add this as a function to ic-nervous-system-common or similar?

Reply:
A small code duplication causes little negative consequences.

Reply:
Closing because I don't see a problem with this. But I'm not against factoring some of these functions out, so long as the added dependency isn't harmful.